src/vsr/replica.zig

```diff
@@ -1020,6 +1083,13 @@ pub fn ReplicaType(
             else
                 self.quorum_replication;
 
+            if (!prepare.ok_from_all_replicas.isSet(message.header.replica)) {
+                self.primary_abdicating = false;
+                if (!prepare.ok_quorum_received) {
+                    self.primary_abdicate_timeout.reset();
```

Comment:
Curiously, this seems to be the single place where `prepare_timeout` and `primary_abdicate_timeout` behave differently. This makes me think this two things: 

* perhaps we can cast `abdiacte_timeout` as just `prepare_timeout` firing several times? Eg, prepare_timeouts.attempts % 4 == 0 or something
* I might be missing something, but it seems we never reset prepare_timeout? This feels like a bug. Suppose cluster is in a steady state with a stable primary, whose pipeline is always non-empty. As we never reset the timeout, we will accumulate attempts. As we do exponential backoffs, our actual prepare tiemout would grow, even everything is ok. So, when we actually do drop a prepare, we'll take a long time to notice that.

Reply:
> perhaps we can cast abdiacte_timeout as just prepare_timeout firing several times? Eg, prepare_timeouts.attempts % 4 == 0 or something

Hmm I'm on the fence here.

Currently we reset whenever any "new" `prepare_ok` arrives, which is a bit more forgiving than using `prepare_timeout.attempts % n == 0`.

e.g. with `attempts % n` and a large pipeline, we might receive a bunch of later prepares while we are blocked (and retrying) an earlier one. If the replicas are successfully acking any prepares, shouldn't that be good enough?

> I might be missing something, but it seems we never reset prepare_timeout?
:+1: I'll add a reset in `on_prepare_ok` when we acquire a quorum (i.e. advance to retrying the next prepare in the pipeline).

Reply:
>Hmm I'm on the fence here.

One tie breaker is that, with a separate timeout, it would be easier to _change_ the logic if for some reason it has to become different.

Reply:
Alright, I'll leave it as-is for now.